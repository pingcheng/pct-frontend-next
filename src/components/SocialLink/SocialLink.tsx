import Link from "next/link";
import styles from "./style.module.css";

export type SocialLinkProps = {
  href: string;
  label: string;
  icon: React.ReactNode;
};

const isExternalLink = (url: string) => {
  return url.startsWith("http://") || url.startsWith("https://");
};

export const SocialLink = ({ href, label, icon }: SocialLinkProps) => {
  if (isExternalLink(href)) {
    return (
      <a
        className={styles.socialLink}
        href={href}
        target="_blank"
        rel="noopener"
      >
        {icon} {label}
      </a>
    );
  }

  return (
    <Link className={styles.socialLink} href={href}>
      {icon} {label}
    </Link>
  );
};
