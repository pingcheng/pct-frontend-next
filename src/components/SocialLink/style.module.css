@reference "../../app/globals.css";

.socialLink {
  @apply cursor-pointer text-gray-400 mr-4 p-2 transition duration-300;
  @apply hover:text-black;
}

.socialLink svg {
  @apply inline-block relative;
  top: -2px;
}