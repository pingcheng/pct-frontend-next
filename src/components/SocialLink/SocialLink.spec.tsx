import { render, screen } from "@testing-library/react";
import { SocialLink } from "./SocialLink";

describe("test <SocialLink />", () => {
  const mockIcon = <span data-testid="icon">Icon</span>;

  it("should match snapshot for external link", () => {
    const { container } = render(
      <SocialLink href="https://example.com" label="Example" icon={mockIcon} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should match snapshot for internal link", () => {
    const { container } = render(
      <SocialLink href="/about" label="About" icon={mockIcon} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should render external link with correct attributes", () => {
    render(
      <SocialLink href="https://example.com" label="Example" icon={mockIcon} />
    );

    const link = screen.getByText(/Example/i);
    expect(link).toHaveAttribute("href", "https://example.com");
    expect(link).toHaveAttribute("target", "_blank");
    expect(link).toHaveAttribute("rel", "noopener");
  });

  it("should render internal link using Next.js Link component", () => {
    render(<SocialLink href="/about" label="About" icon={mockIcon} />);

    const link = screen.getByRole("link");
    expect(link).toHaveAttribute("href", "/about");
    expect(link).not.toHaveAttribute("target");
    expect(link).not.toHaveAttribute("rel");
  });

  it("should render the icon and label correctly", () => {
    render(
      <SocialLink href="https://example.com" label="Example" icon={mockIcon} />
    );

    expect(screen.getByTestId("icon")).toBeInTheDocument();
    expect(screen.getByText("Example")).toBeInTheDocument();
  });

  it("should correctly identify external links", () => {
    render(
      <SocialLink
        href="https://example.com"
        label="HTTPS Example"
        icon={mockIcon}
      />
    );
    expect(screen.getByText(/HTTPS Example/i)).toHaveAttribute(
      "target",
      "_blank"
    );

    render(
      <SocialLink
        href="http://example.org"
        label="HTTP Example"
        icon={mockIcon}
      />
    );
    expect(screen.getByText(/HTTP Example/i)).toHaveAttribute(
      "target",
      "_blank"
    );
  });
});
