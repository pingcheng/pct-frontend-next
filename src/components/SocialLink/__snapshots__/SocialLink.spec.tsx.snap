// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <SocialLink /> should match snapshot for external link 1`] = `
<div>
  <a
    class="socialLink"
    href="https://example.com"
    rel="noopener"
    target="_blank"
  >
    <span
      data-testid="icon"
    >
      Icon
    </span>
     
    Example
  </a>
</div>
`;

exports[`test <SocialLink /> should match snapshot for internal link 1`] = `
<div>
  <a
    class="socialLink"
    href="/about"
  >
    <span
      data-testid="icon"
    >
      Icon
    </span>
     
    About
  </a>
</div>
`;
